### ISSUE TITLE
Nullable boolean tests with [Theory] still add all parameter variants despite [TestCase] attributes

### ISSUE THREAD

**Issue by Gwindalmir** [nunit/nunit3-vs-adapter#678]
NUnit and NUnit3TestAdapter 3.11.0
VS Professional 2017 15.9.17
.NET Core 2.2 and .NET Framework 4.7.2.

Situation:
A test with a bool? argument has all tests variations added, despite explicit [TestCase] values. This not only causes unwanted tests to be generated, but also duplicates where the parameters are the same as the explicit ones.

Observed:
A single test theory with a bool? argument and two [TestCase] attributes generates 5 tests (2 are ignored for being duplicates), leaving a total of 3 runnable tests.

Expected:
Only the two explicit tests should be generated.

The documentation listed at https://github.com/nunit/docs/wiki/Theory-Attribute states:

> For boolean and enum arguments, NUnit can supply the data without any action by the user. All possible values are supplied to the argument. This feature is disabled if the user supplies any values for the argument.

It seems this may not work for nullable values.

The xUnit framework generates the expected test count.

Here's the test that shows the problem:
```
    public class Tests
    {
        [Theory]
        [TestCase(true)]
        [TestCase(null)]
        public void test_theory_should_not_have_false_parameter(bool? unicode)
        {
            Assert.That(unicode, Is.Not.False);
        }
    }
```

Here's the discovery output from Visual Studio:
```
[11/18/2019 5:29:54 PM Informational] ------ Discover test started ------
[11/18/2019 5:29:56 PM Informational] NUnit Adapter 3.11.0.0: Test discovery starting
[11/18/2019 5:29:56 PM Informational] NUnit Adapter 3.11.0.0: Test discovery complete
[11/18/2019 5:29:56 PM Warning] A test with the same name 'Tests.Tests.test_theory_should_not_have_false_parameter(True)' already exists. This test is not added to the test window.
[11/18/2019 5:29:56 PM Warning] A test with the same name 'Tests.Tests.test_theory_should_not_have_false_parameter(null)' already exists. This test is not added to the test window.
[11/18/2019 5:29:56 PM Informational] ========== Discover test finished: 5 found (0:00:02.2650057) ==========
```

Repro case:
[NUnitTheoryTest.zip](https://github.com/nunit/nunit3-vs-adapter/files/3861470/NUnitTheoryTest.zip)

**Comment by mikkelbu** [comment]
Hi @Gwindalmir the issue is not with the adapter, but with either the framework or the documentation. Looking at https://github.com/nunit/nunit/blob/bd45c2e2caae85c530fdd2e9962df382a8fbb863/src/NUnitFramework/framework/Internal/Builders/DatapointProvider.cs the statement you quote is only for data supplied via DatapointAttribute and DatapointSourceAttribute.

The page you found also states 
> In addition to the Datapoint and Datapoints attributes, it is possible to use any of the approaches for supplying data that are recognized on normal parameterized tests. We suggest that this capability not be overused, since it runs counter to the distinction between a test based on examples and a theory. However, it may be useful in order to guarantee that a specific test case is included.

So I think the best solution to this problem would be to improve the documentation rather than possibly breaking peoples code.

**Comment by Gwindalmir** [comment]
> it is possible to use any of the approaches for supplying data that are recognized on normal parameterized tests

This means TestCase, ValueSource, etc, as documented here: https://github.com/nunit/docs/wiki/Parameterized-Tests

So the documentation is at least contradicting itself.

I do understand the necessity to maintain compatibility, but I've noticed NUnit tends to do that to a fault, as I've noticed bad behavior building but never gets fixed because of compatibility.

Anyway, a _safe_ solution could be to add an optional argument to TheoryAttribute to disable automatic test generation.
such as:
```
[Theory(autogenerate = false)]
```
